:bug: Fix race condition: wait for CA bundle ConfigMap before applying CRDs

ccardenosa opened the pull request:
## Problem

The cluster manager controller has a race condition that causes CRDs to be created with an invalid CA bundle, resulting in webhook conversion failures and CRDs not becoming `Established`.

### Symptoms

When deploying OCM (via ACM/MCE on OpenShift), the following cascade of failures occurs:

1. `ClusterManagementAddOn` and `ManagedClusterAddOn` CRDs have `Established: False` with `InvalidCABundle` reason
2. API endpoints are not registered: `"server doesn't have resource type 'clustermanagementaddons'"`
3. MultiClusterHub fails with: `"no matches for kind 'ClusterManagementAddOn' in version 'addon.open-cluster-management.io/v1alpha1'"`

### Root Cause

In `clustermanager_controller.go` (lines 284-296), the code reads the CA bundle from `ca-bundle-configmap`:

```go
caBundle := "placeholder"
configmap, err := n.configMapLister.ConfigMaps(clusterManagerNamespace).Get(helpers.CaBundleConfigmap)
switch {
case errors.IsNotFound(err):
    // do nothing  <-- BUG: silently uses "placeholder"!
case err != nil:
    return err
default:
    // ...
}
```

When the ConfigMap doesn't exist yet (because `certrotation_controller` hasn't created it), the code **silently** continues with `"placeholder"`. This value gets base64-encoded as `cGxhY2Vob2xkZXI=` and injected into the CRDs, causing webhook conversion to fail.

### The Race

1. **certrotation_controller** should create `ca-bundle-configmap` with actual CA certificates
2. **clustermanager_controller** reads this ConfigMap to populate CRD webhook configurations
3. If clustermanager_controller runs first, it uses `"placeholder"` instead of waiting

## Solution

This PR changes the behavior to return an error when the CA bundle ConfigMap is not found, causing the controller to retry until the cert rotation controller has created it:

```go
configmap, err := n.configMapLister.ConfigMaps(clusterManagerNamespace).Get(helpers.CaBundleConfigmap)
if err != nil {
    if errors.IsNotFound(err) {
        klog.V(4).Infof("CA bundle configmap %s/%s not yet available, will retry",
            clusterManagerNamespace, helpers.CaBundleConfigmap)
        return fmt.Errorf("waiting for CA bundle configmap %s/%s to be created by cert rotation controller",
            clusterManagerNamespace, helpers.CaBundleConfigmap)
    }
    return err
}
```

This ensures CRDs are always created with valid CA bundles.

## Testing

- ✅ All existing unit tests pass
- ✅ Added new test `TestSyncDeployWithoutCaBundle` to verify the fix
- ✅ Updated test infrastructure to properly set up CA bundle ConfigMap in test fixtures

## Discovery Context

This issue was discovered during OpenShift CI Prow job runs for ZTP hub deployment:

| Job Run | Date | ACM Version | MCE Version | Status |
|---------|------|-------------|-------------|--------|
| [#2005051399989104640](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-kni-eco-ci-cd-ztp-left-shifting-kpi-ci-4.21-telcov10n-virtualised-single-node-hub-ztp/2005051399989104640) | Dec 27, 2025 | 2.16.0-113 | 2.11.0-142 | ❌ Failed |
| [#2005219283428184064](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-kni-eco-ci-cd-ztp-left-shifting-kpi-ci-4.21-telcov10n-virtualised-single-node-hub-ztp/2005219283428184064) | Dec 28, 2025 | 2.16.0-114 | 2.11.0-143 | ❌ Failed |

Environment: OCP 4.21.0-rc.0

### Evidence from Live Cluster

**CRD with placeholder CA bundle:**
```yaml
spec:
  conversion:
    webhook:
      clientConfig:
        caBundle: cGxhY2Vob2xkZXI=  # Base64 of "placeholder"
```

**CRD Status:**
```yaml
status:
  conditions:
  - type: Established
    status: "False"
    message: The conversion webhook CABundle is invalid
    reason: InvalidCABundle
```

## Changes

- `pkg/operator/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go`
  - Return error when CA bundle ConfigMap is not found (instead of silently using "placeholder")
  - Add validation for empty `ca-bundle.crt` data
  - Improved comments explaining the dependency on cert rotation controller

- `pkg/operator/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller_test.go`
  - Added `newCaBundleConfigMap()` helper to create test ConfigMap
  - Added `newTestControllerWithoutCaBundle()` helper for testing missing ConfigMap scenario
  - Updated `newTestController()` to properly set up CA bundle ConfigMap in informer store
  - Added `TestSyncDeployWithoutCaBundle` test case

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved CA bundle retrieval and validation during cluster manager startup, avoiding deadlocks by ensuring the hub namespace is created earlier.
  * Added retry and clearer diagnostics when the CA bundle is missing or empty.

* **Tests**
  * Added tests for startup behavior with CA bundle present and missing, validating sequencing and error handling.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qiujian16 commented:
/assign @elgnay

elgnay commented on pkg/operator/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go:
```diff
@@ -281,18 +283,32 @@ func (n *clusterManagerController) sync(ctx context.Context, controllerContext f
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("CA bundle configmap %s/%s not yet available, will retry",
+				clusterManagerNamespace, helpers.CaBundleConfigmap)
+			return fmt.Errorf("waiting for CA bundle configmap %s/%s to be created by cert rotation controller",
```
It might be better to requeue it instead of returning an error. 
```
controllerContext.Queue().AddAfter(clusterManagerName, 3 * time.Second)
return nil
```

qiujian16 commented on pkg/operator/operators/clustermanager/controllers/clustermanagercontroller/clustermanager_controller.go:
```diff
@@ -385,6 +401,32 @@ func ensureSAKubeconfigs(ctx context.Context, clusterManagerName, clusterManager
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+	klog.V(4).Infof("Created namespace %s for cert rotation controller", namespace)
```
please use contextual logging

qiujian16 commented:
please rebase to the latest commit, there was a flaky integration test that has been fixed in the latest

ccardenosa commented:
@qiujian16 Done! Rebased to the latest commit (`46de05b`) which includes the flaky integration test fix (#1312).

@elgnay Could you please re-review and re-apply `/lgtm`? The fix is the same as before - just rebased on top of latest main. Thanks!

ccardenosa commented:
## Addressed Review Feedback

I've updated the code based on the review comments:

### 1. Use requeue instead of returning error (@elgnay)

**Before:**
```go
return fmt.Errorf("waiting for CA bundle configmap %s/%s to be created by cert rotation controller",
    clusterManagerNamespace, helpers.CaBundleConfigmap)
```

**After:**
```go
controllerContext.Queue().AddAfter(clusterManagerName, 3*time.Second)
return nil
```

### 2. Use contextual logging (@qiujian16)

**Before:**
```go
klog.V(4).Infof("CA bundle configmap %s/%s not yet available, will retry", ...)
klog.V(4).Infof("Created namespace %s for cert rotation controller", ...)
```

**After:**
```go
logger.V(4).Info("CA bundle configmap not yet available, will retry",
    "namespace", clusterManagerNamespace, "configmap", helpers.CaBundleConfigmap)
logger := klog.FromContext(ctx)
logger.V(4).Info("Created namespace for cert rotation controller", "namespace", namespace)
```

### Test updated

The test was also updated to verify that `sync` returns `nil` (graceful requeue) instead of an error when the CA bundle ConfigMap is missing.

All unit tests pass locally.

@elgnay @qiujian16 - Ready for re-review. Thanks!
